### ISSUE TITLE
Refactor assertion error creation

### ISSUE THREAD

**Issue by joel-costigliola** [joel-costigliola/assertj-core#1296]
#### Summary

There are (too) many ways to create assertion error in the codebase:

- Failures
- AssertionErrorFactory implementation (ShouldBeEqual)
- AssertionErrorCreator

Expected outcome:
- one way to create AssertionError
- deals with building AssertionFailedError when possible or ComparisonFailure.
- handles in one place: 
    - removing assertj stack trace element
    - thread dump

**Comment by arranbartish** [comment]
@joel-costigliola : Loving this project. I'm interested in contributing on this issue. 

As a nice to have it seems like having a way to provide a custom `ValueWrapper` could be nice too. Thoughts?

I'll fork, have a go, and see what you think.

**Comment by joel-costigliola** [comment]
@arranbartish thanks!

What would be the use case for custom `ValueWrapper`? Can you give an example ?

**Comment by arranbartish** [comment]
Certainly @joel-costigliola. Please bear with me. 

# Context

I have a set of complex model objects that models PII and sensitive data, let's say that this is an instance of `Family`. A `Family` is composed of a `Collection<Guardian>` a `Collection<Minor>`, and a `Collection<Pet>`

The production model has no `String` representation of these objects because we don't dare risk that the data escapes the system into unencrypted logs.

This isn't the same problem that we have at test time. We're using the [ObjectMother](https://martinfowler.com/bliki/ObjectMother.html) pattern extensively so that we never test with real data, but always have realistic data.  

# Example

It means that our tests look like this: 


```java

...
import static org.example.family.assertz.Assertions.assertThat;
...

@Test
void will_take_smallest_kid_to_school() {
  final WeekdayFacade sut = new WeekdayFacade()
  final Family expectedFamilyAfterSchoolRun = new FamilyMother().withoutYoungestMinor().build();
  final Family familyBeforeSchoolRun = new FamilyMother().build();
  
  final Family actualFamilyAfterScoolRun = sut.schoolRun(elementary, familyBeforeSchoolRun);
  
  assertThat( actualFamilyAfterScoolRun ).isEqualTo( expectedFamilyAfterSchoolRun );
}
...
```

The custom entry point is providing a custom recursive/reflective Comparator for our model objects so that we're always asserting for value equivalence over equality.  This gives us as little boiler plate in our tests as possible. 

The resulting messages are awesome and unambiguous thanks to a custom instance of a `StandardRepresentation` which looks kind of like this 

```

org.opentest4j.AssertionFailedError: 
Expecting:
 <org.example.family.model.Family@762ef0ea[
  [org.example.family.model.Guardian@1518bde2[
  name="Harry H"
  ssn="sensitive data"
]],
[org.example.family.model.Minor@1818bde2[
  name="Sally"
  age="14"
],
org.example.family.model.Minor@1918bde2[
  name="Fred"
  age="8"
]],
[org.example.family.model.Pet@1318bde2[
  name="Buzz"
  type="Dog"
]]]>
to be equal to:
<org.example.family.model.Family@742ef0ea[
  [org.example.family.model.Guardian@1018fde2[
  name="Harry H"
  ssn="sensitive data"
]],
[org.example.family.model.Minor@1018bde1[
  name="Sally"
  age="14"
]],
[org.example.family.model.Pet@1218bde2[
  name="Buzz"
  type="Dog"
]]]>
when comparing values using AssertableComparator
but was not.
Expected :org.example.family.model.Family@742ef0ea
Actual   :org.example.family.model.Family@762ef0ea
 <Click to see difference>

```

# Limitation

The `AssertionFailedError` bubbles back to the runner and in the IDE you get a convenient `<Click to see difference>`. 

Unfortunately the IDE diff is useless because it seems that the IDE is using the representation in the `ValueWrapper` of the `AssertionFailedError` which is by default a `String.valueOf(value)`. The highlighted difference in the example above becomes `742ef0ea` and `762ef0ea` and ignores the wonderful `StandardRepresentation` mechanisms that AssertJ gives us in the messaging. 

# work arounds

1. Taking the AssertJ message from the console in the IDE and using another diff tool to actually see the difference in an obvious way. 
2. Use the softly mechanisms to assert on each of the Family Attributes independently. The test would need to be updated if we add a `Collection<Visitor>` to a `Family` in the future, or worse, we forget to add the assertion.

# Hopes and Dreams

If I can pull together all the assertion error creation, it seems I might be able to extend the behaviour of the creation to allow one of, or all three of the following:

1. let the system behave exactly as it does passing the `actual` and the `expected`
2. in the event an implementation of a custom `StandardRepresentation` is provided then create the `AssertionFailedError` with a new `AsserJValueWrapper extends ValueWapper` that takes advantage of the custom representation or strategy.
3. allow a specific representation or strategy for the `actual` and `expected` provided independently of the existing `StandardRepresentation` mechanism

Have I understood everything correctly? Thoughts?

**Comment by joel-costigliola** [comment]
Great explanation, I understand the need and I agree it is a good thing to have.

It looks like though there is no way to pass your own `ValueWrapper` to an `AssertionFailedError`, so I guess the trick would be to subclass `AssertionFailedError` and override the getter method to return a `Representation` aware `ValueWrapper`. 

The code that builds `AssertionFailedError` is in `AssertionErrorCreator` (and some bit in `ShouldBeEqual` but that should refactored to use `AssertionErrorCreator`).

**Comment by arranbartish** [comment]
Makes sense. Thanks for the advice. 

I'll go for it in 2 steps. 
1. refactor which I think might also take care of #1335 
2. extend with my hopes and dreams

I'll have a crack at it and get something over for feedback. I think I would like the 2 points to come back as a single PR if it meets what we're expecting. Or would you like a PR for both? My concern is that I want to make sure the refactor (1) is done in a way that enables the extension (2). I want to avoid having to refactor the refactor. 

Would that make sense?

**Comment by arranbartish** [comment]
Would you imagine:
1. refactoring all three `AssertionErrorFactory (ShouldBeEqual)`, `AssertionErrorCreator`, and `Failures` to use something new

-- or --

2. refactoring `AssertionErrorFactory (ShouldBeEqual)` and `Failures`  of them to take advantage of `AssertionErrorCreator`

---------------------------------------

- My initial feeling is to favour `AssertionErrorCreator` (option 2) so that everything goes through that as per expectation 1. **Would this be what you're hoping for?**
- It seems to me that `AssetionErrorCreator` already takes care of expectation 2, with the exception of `ComparisonFailure`. From what I understand we should only be using this when we find it (junit4) on the classpath. **Have I understood this one right?** 
- `Failures` seems to be where the AssertJ stack trace removal lives by making a static call off to `Throwables`. **Would the hope be to consolidate this in the AssertionErrorCreator?** 

1. My aim will be to keep publicly scoped methods unchanged. 
2. Complete the `// TODO deprecate AssertionErrorFactory` in `AssetionErrorCreator` and add the `@deprecated` marker to `AssertionErrorFactory`

If I am successful, my expectation would be that I meet the definition of "refactor":
1. behaviour is exactly the same
2. no indication to a consumer that anything has changed (The exception to this would be the `@deprecated` marker)
3. none of the existing tests would change

**Does this seem fair?**

**Comment by joel-costigliola** [comment]
That seems ok to me, I'm just restating the refactoring goals so that it is :
* deprecate `AssertionErrorFactory` and make sure it is not used anymore (i.e. replacing `failure(AssertionInfo info, AssertionErrorFactory factory)` by `failure(AssertionInfo info, ShouldBeEqual shouldBeEqual)`
* refactor `ShouldBeEqual` to use `AssertionErrorCreator` to build assertionFailedError
* make `AssertionErrorCreator.assertionFailedError` to call  `Failures.instance().removeAssertJRelatedElementsFromStackTraceIfNeeded(assertionError);` to clean up stack trace if needed (`AssertionErrorCreator` should have been doing that).

Let's create a first PR for the refactoring to keep the changes small and easier to review. Once this is done, I'll evaluate the opportunity to do further refactoring but I'll create another issue for that (for example let `AssertionErrorCreator` create `ComparisonFailure` when possible).

In my mind this does not take care of #1335 which was meant to spread the use of `AssertionFailedError` in assertions like `assertHasMessageContaining` (a Throwable assertion), I don't have a complete list in mind, it was more of reminder to think about it ;-)

The support of `Representation` in `AssertionFailedError` should come afterwards in a different PR, don't hesitate to create a PR even if the work is not complete if you need some feedback from me.
